### PR TITLE
Ocaml benchmarks

### PR DESCRIPTION
**Pull request by gasche**
This is a bit of a monster PR as I am commuting out, sorry.

A lot of this is scripts I use to produce the LaTeX tables and plots automatically. But the rest is also interesting:
- zipper implementation of replay non-determinism, in OCaml and SML (including array/vector versions to remove the (List.nth) inherent in choose+replay)
- rename the SML files to follow the OCaml scheme (to have a unified description in the benchmark sections, if we decide to show both families of numbers)

The PR contains the `result.data` that I generated on my machine, which suffices to reproduce the `.tex` parts that I committed in the paper -- and change them if we tune the automated output again, etc.

### CODE REVIEW

File: benchmarks/nqueens/ocaml/thermometers_refactored.ml [hunk 1]
```diff
@@ -49,10 +49,10 @@ module Control (A : Answer) : CONTROL with type ans = A.ans = struct
      let res = try f () with Done x -> x in
      match pop reset_stack with
      | None -> assert false
-     | Some (f', rec_stack', rep_stack') ->
-       cur_fun := f';
-       record_stack := rec_stack';
-       replay_stack := rep_stack';
+     | Some (future, record, replay) ->
```
Comment (jkoppel) [inline benchmarks/nqueens/ocaml/thermometers_refactored.ml]: Cute, but the cur_fun contains both the future and the past.

File: benchmarks/nqueens/run_benchmark.rb [hunk 2]
```diff
@@ -15,13 +15,18 @@
...
 LOWERBOUND.upto(UPPERBOUND) do |n|
-  tot = 0.0
-  TRIALS.times { tot += `sml #{file} #{n}`.lines[-1].to_i }
-  avg = 1 + tot / TRIALS
-  puts "point\t#{n}\t#{avg}"
+  best = 0
+  TRIALS.times {
+    unless best > 60*1000*2 then
+      run = `timeout 10m sml #{file} #{n}`.lines[-1].to_i
+      if best = 0 || run < best then best = run end
```
Comment (jkoppel) [inline benchmarks/nqueens/run_benchmark.rb]: Did you mean "if best == 0"?

File: benchmarks/nqueens/run_benchmark_time.rb [hunk 3]
```diff
@@ -19,20 +19,23 @@
...
-    #p s
-    m = s.match(/user\s+(\d+)m(\d+.\d+)s/)
-    tot += (m[1].to_i * 60 + m[2].to_f) * 1000
+    unless best > 60*1000*2 then
+       _, stderr, _ = Open3.capture3("bash -c \"time #{cmd} #{n}\"")
+      s = stderr.lines.join
+      #p s
+      m = s.match(/user\s+(\d+)m(\d+.\d+)s/)
+      run = (m[1].to_i * 60 + m[2].to_f) * 1000
+      if best = 0 || run < best then best = run end
```
Comment (jkoppel) [inline benchmarks/nqueens/run_benchmark_time.rb]: Ditto above

**Comment by jkoppel** [comment]
I skimmed over everything. It looks good...

...except for the "if best = 0" issue, which looks quite concerning. What's going on there?

**Comment by gasche** [comment]
Indeed; I tried to get the "best out of N" instead of "average over N" that people find more stable for benchmarks (I'm no expert, but was told that the outliers are more slow-outliers, so "best" filters them out). Right now the code only takes the first (or last?) run, I suppose. I will rerun the benchmarks with the fixed code.

**Comment by jkoppel** [comment]
Glad we caught that.

**Pull request merged**